### ISSUE TITLE
ompr next steps

### ISSUE THREAD

**Issue by dirkschumacher** [dirkschumacher/ompr#283]
Hi all,

I have been thinking about how to advance modelling in R and recently I had some energy again to spend on optimization. 

One idea is to have a central, mutable solver object instead of an immutable model object (julia JuMP approach), as it enables communicating directly with the solver. It is then possible to implement things like quadratic programs, callbacks or very solver specific stuff.

I started implementing and testing these ideas in a separate package this week: https://github.com/dirkschumacher/rmpk The package shares most of the API with `ompr` so it feels quite familiar. Another goal is also to make the package so fast that you can use it within a package as well as in interactive analyses.

There are three options I have, once the experimental package is complete and I am satisfied with the result.

1) Release this as an independent package and leaving ompr as it is
2) Merge this with ompr as a third "model backend" while deprecating the other two.
3) Using the new code in the `ompr` package thus breaking all previous code. (But it would not be too hard to fix any broken code as the API is roughly the same)

Any input is very welcome.

PS: I cannot give any timeline on the new development as I am just doing that after work when I have some energy.

**Comment by devu123** [comment]
Can you elaborate on `making the package fast' ? If you have plans to make it utilize multiple threads / easily deploy solving of models on a remote server, I would lean towards option 1. If not, option 2 seems best. Is their any benefit to option 3?

**Comment by prof-anderson** [comment]
Interesting.  I like the new way of building models.  Not sure if prefer it but if has major benefits, I'm all in.

Architecturally, it seems to be a way of overcoming the differences between solvers and supporting quadratic models which may both be more difficult given ompr's current architecture.  Presumably it would not then require parsing or interpreting the status flags for each solver in terms of differentiating between stopped prior to proven optimal vs. infeasible.

The models that I've been examining and building have not taken long for ompr to build with on the order of a thousand variables and/or constraints.  Solver performance has been more variable and better control options/feedback would be nice.  

I agree with devu123 in that I'm not seeing an obvious benefit for option 3.  I would prefer to know that my old models still work on possibly an orphaned package and then be able to rewrite them to the new package.  

This seems like it might be harder to run a model on different solvers since the solver is specified as part of the model rather than at time of solving.  (Perhaps the solver can be changed but I could envision that being tricky too.)

My wish list from highest  priority.  Not all of these related to or solveable by the new architecture but I thought that I would at least mention them.  If anyone has suggestions for 

1. Better termination of models being solved (In Rstudio and Windows, I can't generally stop a long run without having to kill the whole task in task manager.  Perhaps this is an Rstudio issue.)
2. A fix to solution status being misinterpreted as infeasible for prematurely terminated runs and for unbounded problems.
3. Nonlinear solving.  Would be nice to have it within the same general structrure/framework/package.(quadratic of course, but perhaps someday more generalized nonlinear?)  
4. Model building speed (hasn't been an issue for me so far, mostly due to solver issues keeping me from doing big stuff)
5. Better control and selection of solvers, perhaps this is more of an ROI issue.  For example, was hoping to try gurobi but they don't seem to support ROI directly and the third party roi.plugin.gurobi seems to be problematic.

I may try doing some experimentation with the new package over the holidays.

Great job!  I've been using ompr a lot for a class that I teach.  I've had some interest from people at other universities in adopting it as well.

**Comment by dirkschumacher** [comment]
Thanks for your comments!

So I am making progress on the [rmpk](https://github.com/dirkschumacher/rmpk) package.

So far it has the following improvements:
* Faster modelling speed
* Quadratic and Linear Programming supported
* Callbacks for solvers. E.g. if you want to use row generation to solve [TSP problems](https://dirkschumacher.github.io/rmpk/articles/tsp-subtour.html)
* Better control over termination status. E.g. a termination after a time limit.
* Ships with optional ROI built in
* Character indexes for variables
* Can be used within other packages
* No C++
* Architecture supports many solver specific features not currently possible, like Special Ordered Sets, callbacks, etc.

I am still not sure how to integrate that with ompr in a backwards compatible way. Especially supporting pipes is tricky - not impossible, but tricky.

I am not sure if general non-linear programming will ever be supported (it would be cool though). For that it would need an even more abstract architecture that is currently out of scope given my very limited free time I can devote to open-source software.

**Comment by prof-anderson** [comment]
It looks like great progress on rmpk!  Other than piping, it seems to have most features of ompr already and a lot more.  I don't mind losing piping -having a piped set of model building commands sometimes makes it harder to identify where my errors.

You say "better control over termination status" - does this mean rmpk handles ROI's inconsisent solver status codes different from ompr?  

Tim
